pkg/ccl/sqlproxyccl/forwarder.go

```diff
@@ -0,0 +1,158 @@
+
+	// errChan is a buffered channel that contains the first forwarder error.
+	// This channel may receive nil errors.
+	errChan chan error
+
+	mu struct {
+		syncutil.Mutex
+
+		// closed is set to true whenever the forwarder is closed explicitly
+		// through Close, or when any of its main goroutines has terminated,
+		// whichever that happens first.
+		//
+		// A new forwarder instance will have to be recreated if one wants to
+		// reuse the same pair of connections.
+		closed bool
+
+		// started is set to true once Run has been invoked on the forwarder.
+		// This is a safeguard to prevent callers from starting the forwarding
+		// process twice. This will never be set back to false.
+		started bool
+	}
+}
+
+// newForwarder returns a new instance of forwarder. When this is called, it
+// is expected that the caller passes ownerships of conn and crdbConn to the
+// forwarder, which implies that these connections may be closed by the
+// forwarder if necessary, but this is not guaranteed. The caller should still
+// attempt to cleanup the connections.
+func newForwarder(ctx context.Context, conn net.Conn, crdbConn net.Conn) *forwarder {
+	ctx, cancelFn := context.WithCancel(ctx)
+
+	return &forwarder{
+		ctx:       ctx,
+		ctxCancel: cancelFn,
+		conn:      conn,
+		crdbConn:  crdbConn,
+		errChan:   make(chan error, 1),
+	}
+}
+
+// Run starts the forwarding process for the associated connections, and can
+// only be called once throughout the lifetime of the forwarder instance.
+//
+// All goroutines spun up must check on f.ctx to prevent leaks, if possible. If
+// there was an error within the goroutines, the forwarder will be closed, and
+// the first error can be found in f.errChan.
+//
+// This returns ErrForwarderClosed if the forwarder has been closed, and
+// ErrForwarderStarted if the forwarder has already been started.
+func (f *forwarder) Run() error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if f.mu.closed {
+		return ErrForwarderClosed
+	}
+	if f.mu.started {
+		return ErrForwarderStarted
+	}
+
+	go func() {
+		defer f.Close()
+
+		// Block until context is done.
+		<-f.ctx.Done()
+
+		// Closing the connection will terminate the goroutine below.
+		// This will be here temporarily because the only way to unblock
+		// io.Copy is to close one of the ends. Once we replace io.Copy
+		// with the interceptors, we could use ctx directly, and no longer
+		// need this goroutine.
+		f.crdbConn.Close()
+	}()
+
+	// Copy all pgwire messages from frontend to backend connection until we
+	// encounter an error or shutdown signal.
+	go func() {
+		defer f.Close()
+
+		err := ConnectionCopy(f.crdbConn, f.conn)
+		select {
+		case f.errChan <- err: /* error reported */
+		default: /* the channel already contains an error */
+		}
+	}()
+
+	f.mu.started = true
+	return nil
+}
+
+// Close closes the forwarder, and stops the forwarding process. This is
+// idempotent.
+func (f *forwarder) Close() {
+	f.mu.Lock()
```

Comment:
What do you think of replacing Close() with context cancellation? Everywhere we check for .closed, it could instead check ctx.Err(). That allows us to drop the mutex and reduces the amount of code related to clean up.